src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs

```diff
@@ -492,6 +506,20 @@ private void ProcessEvents()
                                 }
                             }
                             expandedName = associatedDirectoryEntry.GetPath(true, nextEvent.name);
+
+                            // inotify picks up changes to symlinks and reports them as file changes, which breaks the contract. We need
```

Comment:
Good catch, removing this part which also should resolve your comment below about the extra alloc and syscall